Support __index and __newindex metamethods

mpeterv opened the pull request:
Hi again! This PR adds support for inheritable __index and __newindex metamethods that respect inheritance chain, that is, they are called only when the key is missing not only in the instance, but also in the class and all parent classes. Both functions and tables are accepted as values for __index and __newindex. New tests included.

kikito commented:
Wow, thanks a lot for the PR. It is however a lot of code for what I consider a very edge-case use. My first instict is not to accept this pull request just because of that. However, let me go over it again this weekend.

mpeterv commented:
It would be useful for implementing properties and container classess. pl.class doesn't support this, which leads to various issues with its Map classes, for instance stevedonovan/Penlight#173.

Now that I think of it, it doesn't make sense not to call custom __newindex when the key is found in the class. For __index wrapping is necessary to preserve default indexing behavior (fallback to class dictionary), but there is no such behavior when setting a key. So I think the wrapper for __newindex only restricts users and should be removed, even if there wasn't too much code.

kikito commented:
> It would be useful for implementing properties and container classess

I know. My problem is that I think most people using middleclass don't need to do that. If I had to estimate I would say maybe 5% of the users need those things.

While this PR works, the library goes from ~200 LOC to ~300 LOC. There's a 50% increase in code size to benefit a 5% of the users. Looking at these numbers alone makes me think it is not a good tradeoff.

*BUT* I have an idea to make everything smaller. I don't know if it's possible, but it is worth investigating. Let me explain: right now, thanks to your previous PR, we handle instance methods one way (let's call this way "the A way"), and metamethods in a different one ("the B way").

The idea is: wouldn't it be possible to handle *all methods* with the B way? If my thinking is correct, we would not have to "handle metamethods especially" any more. Some `ifs` would disappear. The `_metamethods` table would be gone. And the code to handle the "A way", would be gone. If it works, the instances would be faster, too (in a hierarchy of classes X parent of Y parent of Z, an instance of Z would only have to look up for methods on Z, not needing to "go up" X and Y any more).

Once that is done, the only exceptions remaining would be `__index` (and maybe `__newindex`. I am not sure about if after your comment). That would be the, "exception" now, or the "C way".

There are several unknowns here. As I said, I don't know for sure if it's possible to go "all B". If it works, middleclass would probably go under 200 LOC, and probably faster, which I would appreciate in any case. Once this is done, I don't know how much extra code would be needed for adding C way, but I suspect the end result might still fall under 200 LOC.

I might give this idea a try this weekend, but feel free to send pull requests (or change your current one) if you want!

mpeterv commented:
> While this PR works, the library goes from ~200 LOC to ~300 LOC.

It was only +50 lines (half the changes were new tests). I've removed __newindex wrapping (not only to reduce size, I actually think it's better that way!) and reverted a formatting change, so it went down to +30 lines.

> The idea is: wouldn't it be possible to handle all methods with the B way?

I think it should be possible. A minor issue is that if we supported `__index` and used the same table as instance metatable and as method dictionary, then `aClass.__index` would return `__index` wrapper instead of function set by user. We would also need a way to see if `__index` was defined in a class or merely inherited. Other than that, it should work. Speed of method look-up would improve slightly, at the cost of slower method definition, but I don't think it would ever matter in a real scenario.

Potentially a larger issue is that there isn't much code supporting the "A way". There is one __index set to fallback from class to its parent, and Lua handles the rest itself. I'll see what I can do though.

mpeterv commented:
@kikito so what do you think about the PR? It got smaller.

kikito commented:
Hi, sorry, the last two weeks have been a bit ... eccentric.

I gave it a look. It is a bit complex for my taste as it is. I tried to simplify by removing the "dual way of resolving things" that I mentioned eariler. But I got lost in the process - a bit of too much indirection.

I will try to give this another shot.

kikito commented:
Hi there,

Since this PR breaks some backwards-compatibility (if you manipulated `__instanceDict`, as I did in [stateful.lua](https://github.com/kikito/stateful.lua) ) I have decided to move this to a new major release (4.0). You can see what I have done so far on this branch: https://github.com/kikito/middleclass/tree/v4.0

I still haven't released it because I want to see if I can unify the lookup of methods and metamethods as I explained above. If I am not able, your PR will be part of 4.0.

For now, I hope it's ok If I close this issue.

kikito commented:
Hi, I just merged an (extensively modified) version of your PR into master and released v4.0. Thanks a lot and happy new year!

mpeterv commented:
Hi, thanks, I'll check it out! Happy new year!
